stdlib: add `Random.int_in_range` and similar functions

gmevel opened the pull request:
This PR adds the following functions to `Random`:

```ocaml

val int_in_range : min:int -> max:int -> int
(** [Random.int_in_range ~min ~max] returns a random integer
    between [min] and [max] (both bounds inclusive).
    [min] must be less than or equal to [max].
*)

val int32_in_range : min:int32 -> max:int32 -> int32
val int64_in_range : min:int64 -> max:int64 -> int64
val nativeint_in_range : min:nativeint -> max:nativeint -> nativeint
(* similar *)
```

and their counterpart in the `Random.State` sub-module:

```ocaml
  val int_in_range : t -> min:int -> max:int -> int
  val int32_in_range : t -> min:int32 -> max:int32 -> int32
  val nativeint_in_range : t -> min:nativeint -> max:nativeint -> nativeint
  val int64_in_range : t -> min:int64 -> max:int64 -> int64
```

## Motivation

The already-existing function for drawing integers uniformly in an interval, namely `Random.full_int`, has limited functionality:

(1) expressiveness: it does not support the full range of integers (it can only draw integers between 0 and `max_int`−1), so we cannot easily use it to emulate drawing from an arbitrary interval;

(2) code clarity: even when its range suffices, it must be used in a convoluted way if wanting to draw from an interval whose lower bound is not zero; namely, the code pattern `min + Random.full_int (max - min + 1)`, which is not terribly explicit, prone to errors (it is easy to forget the `+1`); worse, this code is broken in the general case because `max - min + 1` may overflow [this is a re-statement of point (1)].

Hence these new functions are useful. and they are not trivial to implement, (and indeed [the implementation in Containers][containers-bug] has the mentioned bug, for instance).

[containers-bug]: https://github.com/c-cube/ocaml-containers/blob/1a23731/src/core/CCRandom.ml#L49

## Interface

The proposed interface tries to be as abundantly explicit as possible about what the two integer arguments are: inclusive bounds (by contrast with exclusive bounds, or a length for the second argument, or something else yet…). This goes with the name "`int_in_range`" and labeled arguments "`~min ~max`".

Related functions in third-party libraries:
  - Containers has [`int_range : int -> int -> State.t -> int`][doc-containers].
  - Base has [`int_incl : int -> int -> int`][doc-base] and similar for `int32`, `int64`, `nativeint` (it also has `float_range : float -> float -> float` where the upper bound is *exclusive*).

[doc-containers]: https://c-cube.github.io/ocaml-containers/3.12/containers/CCRandom/index.html#val-int_range
[doc-base]: https://ocaml.org/p/base/v0.16.2/doc/Base/Random/index.html#val-int_incl

## Future work

An analogous function for `float` should be provided, but this PR does not tackle it. This might be as simple as `min +. float state (max -. min)` (as Base does it), but I do not feel bold enough to venture into the floating-point land (the subtraction scares me). Also, the documentation states that the bounds are inclusive, but the implementation makes it clear that both zero and the upper bound are in fact exclusive (unless `bound = 0.0`).

## Side comments

I also added one explanatory comment in the internal implementation of `Random`.

Regarding said implementation: why isn’t `intaux` replaced with `int63aux` entirely? In 32-bit OCaml, both functions are identical (so "`int63aux`" is a misnomer). In 64-bit OCaml, `intaux` produces the same thing as `int63aux`, only with a restricted input range, and (if I get it right) a higher probability of re-drawing. `intaux` does not even spare some random bits, it draws 64 bits of randomness anyway.

gmevel committed:
Random: add comment

Explain a non-obvious bit of the implementation.

gmevel committed:
add Random.int_in_range and related

New functions:
- `Random.int_in_range`
- `Random.int32_in_range`
- `Random.int64_in_range`
- `Random.nativeint_in_range`
- `Random.State.int_in_range`
- `Random.State.int32_in_range`
- `Random.State.int64_in_range`
- `Random.State.nativeint_in_range`

gmevel commented:
Answering my own question:

> why isn’t `intaux` replaced with `int63aux` entirely?

I guess this is for backwards compatibility, to preserve pseudo-random sequences with a given seed? In that case, the new function `int_in_range` does not need to follow that logic which I dumbly mimicked from `full_int`, of switching between `intaux` and `int63aux` in function of the size of the arguments.

Waiting for confirmation. If my new understanding is correct, I believe it would be worthwhile to clarify the implementation by adding a small comment and by renaming `intaux` to `int31aux` and `int63aux` to `full_int_aux`.

gasche commented:
I believe that the reason for `intaux` is for 32bit and 64bit generators to behave in the same way (on 32bit inputs), improving portability for computations that run on both kinds of machines, at little performance cost (compared to performing all computations on Int64 numbers, which would be fairly costly on 32bit machines).

gasche commented on stdlib/random.mli:
```diff
@@ -63,18 +63,46 @@ val full_int : int -> int
+val int_in_range : min:int -> max:int -> int
+(** [Random.int_in_range ~min ~max] returns a random integer
+    between [min] and [max] (both bounds inclusive).
+    [min] must be less than or equal to [max].
```
I understand that both `min` and `max` could be negative. This may be natural, but it different from all other integer generators, so it may deserve an explicit mention.

gasche replied:
I would find `between [min] (included) and [max] (included)` simpler to read than the current wording `between [min] and [max] (both bounds inclusive)`.

gasche replied:
It would be nice to be explicit about exceptions:

```
@raise Invalid_argument if [min > max]
```

xavierleroy replied:
Can you please add the `@raise` ?

gmevel commented:
Oh, that makes sense. In that case `int_in_range` should stick to that logic, I guess. This definitively deserves a comment.

gasche commented:
More on this question of intaux vs. int63aux:
- We changed our PRNG algorithm for 5.0 to support splitting.
- If I remember correctly, the previous OCaml PRNG was tuned to generate exactly 30 random bits on each state update, so `Random.int` would fail (and still does) with an exception if asking for higher numbers. Note that even on 32bit systems, 2^30 (about one billion) is strictly smaller than `max_int` which is 2^31-1 (about two billions) (let's call this `max_int_32`), so users could observe a failure.
- `Random.int` has the 32-64-compatibility property: it produces the same output on 32bit and 64bit machines on 32bit-representable inputs, which is (for this function) all non-failing inputs.
- `Random.full_int` was added to generate larger numbers in #9489 , before the change or PRNG algorithm ; the implementation is not simple and the discussion was very long! In particular, its implementation was careful to provide the 32-64-compatibility property for bounds above 2^30 and below `max_int_32`, on which `Random.int` still fails.
- The implementation was modified and much simplified when we migrated to the new PRNG. `Random.int` is still limited to bounds below 2^30, but on the other hand for `full_int` I don't know if the 32-64-compatibility property still holds today -- this is not obvious from the implementation, so a question for @xavierleroy.

It would be nice if `int_int_range` also satisfied the 32-64-compatibility property. I suspect that it does not satisfy it today: it is obviously compatible when `0 < span <= 2^30`, but there are cases of 32-bit `min, max` inputs that give `span > 2^30` on all machines, and also some cases that have `span <= 0` on 32bit machines and `span > 2^30` on 64bit machines.

gasche commented:
It would also be nice to have the following properties:

```
forall (0 < n <= 2^30), Random.int n = Random.int_in_range ~min:0 ~max:(n-1)
forall (0 < n),         Random.full_int n = Random.int_in_range ~min:0 ~max:(n-1)
```

I believe that the implementation as proposed does satisfy these properties: with `~min:0 ~max:(n-1)` we have `span = n` and the result is exactly `Random.full_int n`. (Same for int32, int64, nativeint.)

Documenting them somewhere may be useful.

gasche reviewed (commented):
I agree that these functions are useful to have, and I find the naming choices tasteful.

I believe that the implementations are correct (including the conditional usage of `intaux`), except possibly for corner cases of 32-64-compatibility of `int_in_range` -- which I would prefer to see fixed.

In other words, I am heading towards approval. Stdlib PRs need two maintainer approvals, and PRNG stuff should get an opinion from @xavierleroy in any case.

gmevel committed:
Random/doc: improve doc

- move the note about @before 5.0 to the top of the module,
  because it affects all functions rather than `bits` specifically
- document all the Invalid_argument raised

xavierleroy commented:
> The implementation was modified and much simplified when we migrated to the new PRNG. Random.int is still limited to bounds below 2^30, but on the other hand for full_int I don't know if the 32-64-compatibility property still holds today 

I see only one documented compatibility property:
```
If [bound] is less than 2{^30}, [Random.full_int bound] is equal to {!Random.int}[ bound].
```
It trivially holds because `full_int` calls `intaux` if the bound is less than 2^30.

Re-reading #9489, I see there was a special case for JS-of-OCaml and its 32-bit integers that disappeared in OCaml 5.  Oh, well, nobody noticed :-)  This will have to be discussed separately.

xavierleroy commented on stdlib/random.ml:
```diff
@@ -132,8 +133,18 @@ module State = struct
+    (* We start by drawing any representable non-negative integer. *)
     let r = Int64.to_int (next s) land max_int in
     let v = r mod n in
+    (* For uniform distribution of the result between 0 included and [n]
+     * excluded, the random number [r] must have been drawn uniformly in
+     * an interval whose length is a multiple of [n]. So the test below
+     * checks whether [r < k*n], where [k*n] is the greatest multiple of
+     * [n] such that [k*n - 1] is in the interval of representable
+     * integers. If failing that, any interval whose length is
+     * a multiple of [n] and which contains [r] has values that are not
+     * representable, hence, drawing was not uniform; in that case, we
+     * re-draw. *)
```
You're spoiling the mystery with this comment :-)  I enjoyed spending 15 minutes re-doing the math every time I looked at this code :-)

The "If failing that" part is not very clear to me, and I wish you would mention the magic words: "rejection sampling".  I'll try to make a more constructive suggestion later.

gmevel replied:
Ok, I am not familiar with that terminology, so I let you amend that comment.

xavierleroy replied:
https://en.wikipedia.org/wiki/Rejection_sampling

Octachron replied:
Maybe

> For uniform distribution of the result between [0] included and [n] excluded, the random number [r] must have been drawn uniformly in an interval whose length is a multiple of [n]. To achieve this we use rejection sampling on the greatest interval [ [0, k * n[ ] that fits representable integers by rejecting samples outside of this interval with the test below.

?

gmevel replied:
Drawing inspiration from @Octachron I reworded the comment as follows:

```ocaml
    (* For uniform distribution of the result between 0 included and [n]
     * excluded, the random number [r] must have been drawn uniformly in
     * an interval whose length is a multiple of [n]. To achieve this,
     * we use rejection sampling on the greatest interval [ [0, k*n-1] ]
     * that fits in representable integers, that is, we reject the
     * sample if it falls outside of this interval, and draw again.
     * This is what the test below does (while carefuly avoiding
     * overflows and sparing a division [max_int / n]). *)
```

Are you satisfied with the new wording?

xavierleroy replied:
OK with me, thanks!

gmevel commented:
Thanks for the background! I agree that the 32-64-compatibility is desirable, and that the current implementation does not satisfy it. I will add a test to check that.

About what you said:

> Note that even on 32bit systems, 2^30 (about one billion) is strictly smaller than `max_int` which is 2^31-1 (about two billions) (let's call this `max_int_32`), so users could observe a failure.

Unless I’m misunderstanding something, `Int.max_int` on 32-bit OCaml *is* 2^30-1, not 2^31-1 (because then ints have 31 bits and are signed). So users on 32-bit OCaml could not observe the failure.

> [the implementation of `Random.full_int`] was careful to provide the 32-64-compatibility property for bounds above 2^30 and below `max_int_32`, on which `Random.int` still fails.

I don’t follow here: are you saying that 32-64-compatibility implies to produce the same result on 32-bit and 64-bit platforms when 2^30 <= `bound` <= 2^31-1? Such values are not representable with the type `int` on 32-bit platforms, so I don’t understand that point about `full_int`. However it makes sense for `int_in_range` because, even on 32-bit platforms, the `span` of the requested range (i.e. `max-min+1`) can be such that 2^30 <= `span` <= 2^31.

I agree with the other compatibility properties you mention. I will document them and try to add tests for them as well.

xavierleroy commented on stdlib/random.ml:
```diff
@@ -144,6 +155,28 @@ module State = struct
+      (* We use [int_in_range_aux] only if [max - min + 1] overflows;
+       * then, the interval between [min] and [max] covers at least half
+       * of the representable integers, so that it converges quickly
+       * (the probability of re-drawing is at most 1/2). *)
```
Nice trick!

xavierleroy reviewed (approved):
Looks good to me.  I usually prefer semi-open intervals `[min, max+1)` to closed intervals `[min, max]`, but I agree that it's nice to be able to take `max = max_int`.

xavierleroy commented:
Trying to clarify the discussion: the "32-64-compatibility" property is ill-defined.  There are THREE sizes for OCaml's type `int`:
- 31 bits for plain OCaml on a 32-bit platform
- 32 bits with JS-of-OCaml
- 63 bits for plain OCaml on a 64-bit platform.

The fancy footwork in #9489 was to guarantee consistent results for `full_int` between JS-of-OCaml and OCaml-on-a-64-bit-platform in the case `0x3FFF_FFFF < bound <= 0x7FFF_FFFF`.  But that's been lost in OCaml 5 and will have to be rediscussed elsewhere.

xavierleroy commented:
Concerning the equalities `Random.int n = Random.int_in_range ~min:0 ~max:(n-1)` et al, I don't think they should be guaranteed, even if they happen to hold in the current implementation.  Two different ways of producing random numbers have no reason to produce the same random numbers.

gmevel committed:
Random: more comments

gmevel committed:
Random: make int_in_range yield same output on 32-bit and 64-bit

This was already the case of `Random.full_int`.
Comments have been added in the implementation to clarify that point.

A test has been added, which covers `full_int` and `int_in_range`.

gmevel committed:
Random: rename internal functions

- `intaux` -> `int31aux` (because this one is for 31-bit integers)
- `int63aux` -> `full_int_aux` (because this one is for any value of
  type `int`, regardless of the int size)

gmevel committed:
Random/doc: document 32-64-compat of full_int, int_in_range

gmevel commented:
I rebased, addressed all minor remarks but one, and in follow-up commits I addressed the "32-64-compatibility" question (the fact that, for 31-bit input, `Random.full_int` and `Random.int_in_range` yield identical output regardless of int size):

  - enforced it for `Random.int_in_range`;
    + the code is straightforward and I believe it is correct, but perhaps there is a clever way to reduce branching;
  - added clarifying comments for `Random.full_int`, and I also took the liberty of renaming internal functions as I suggested (`intaux -> int31aux`, `int63aux -> full_int_aux`);
  - documented it for both `Random.full_int` and `Random.int_in_range`;
  - added a test for both `Random.full_int` and `Random.int_in_range`.

No effort has been made to extend this compatibility to 32-bit integers (as found in JavaScript environments). Future work, as Xavier said.

Regarding the other compatibility property (the fact that `full_int n = int_in_range ~min:0 ~max:(n-1)`, and similarly for other integer types): I mentioned it in an implementation comment. Since there is no consensus yet on whether it should be guaranteed, I did not document it, nor wrote a test for it.

On my machine, I have one build failure that I cannot understand:

```text
  LINKC ocamlc
File "_none_", line 1:
Error: Generated bytecode executable "ocamlc" cannot be used on a 32-bit platform
```

gmevel committed:
Random: reword comment about rejection sampling

xavierleroy commented on stdlib/random.ml:
```diff
@@ -114,35 +114,106 @@ module State = struct
+     We must have [min <= max], and [min] and [max] must fit in 31 bits.
+     This function is designed so that it yields the same output on
+     32-bit and 64-bit platforms. *)
+  let rec int31_in_range_aux s ~min ~max =
+    let r = Int64.to_int (next s) land 0x7FFF_FFFF in
+    if r < min || r > max then int31_in_range_aux s ~min ~max else r
```
This is wrong on a 32-bit platform because 0x7FFF_FFFF cannot be represented as an int.  (That's why you get a link-time error, even if you're linking on a 64-bit host.)

This is wrong on a 64-bit platform because `r` is always >= 0 (as a consequence of the `land`), while `min` can be negative.

gmevel replied:
Regarding 32-bit platforms: I’m confused. I don’t have a 32-bit system at hand, but I verified on my 64-bit system that OCaml (toplevel/compiler) accepts the equivalent constant `0x7FFF_FFFF_FFFF_FFFF` and maps it to `-1`. Is the behavior different for 32-bit OCaml?

Regarding 64-bit platforms: indeed, my mistake.

New plan to reduce the sample to 31 bits: keep the 30 lowest bits, and replicate the sign bit in the remaining higher bits.

```ocaml
    (* 0x3FFF_FFF = 2^30 - 1 has its 30 lowest bits equal to one.
     * -0x4000_000 = -2^30 has all bits equal to one except the 30 lowest bits which are equal to zero.
     *)
    let r = Int64.to_int (next s) in
    let r = (if r >= 0 then r land 0x3FFF_FFFF else (r land 0x3FFF_FFFF) lor -0x4000_0000) in
```

or, without branching:

```ocaml
    let r = Int64.to_int (next s) in
    let s = r asr Sys.int_size in (* this replicates the sign bit of [r] to all bits *)
    let r = (s land -0x4000_0000) lor (r land 0x3FFF_FFFF) in
```

Does that seem correct to you?

xavierleroy replied:
1-  OCaml on a 32-bit platform will indeed accept 0x7FFF_FFFF and read it as -1.  However, if you compile to bytecode on a 64-bit platform, 0x7FFF_FFFF reads as 2^31-1, and that's the constant that would be put in the bytecode executable, making it impossible to run it on a 32-bit platform.  That's what the "Generated bytecode executable "ocamlc" cannot be used on a 32-bit platform" error means.

2- Sign extension of the low 31 bits sound like a good idea, but can be achieved much more easily:
```
let d = Sys.int_size - 31 in (r lsl d) asr d
```

gmevel replied:
1. Thanks for the explanation!
2. That’s much better indeed. By contrast with my version, yours also has the good taste of computing, in fact, the signed remainder modulo 2^31 of its input, even when `int`s have more than 31 bits (because yours replicate the 31th bit whereas mine replicate the highest bit); so it makes sense to isolate this code in a function `int31_of_int`. Regarding the constant `d`, do I need to write it in some way so that the compiler optimizes it to an immediate constant operand for the shift?

gmevel replied:
I applied the suggested fix, and I also added (numerous) chi2 tests.

[For future readers (since the commit has now disappeared): the faulty code was:
```ocaml
  let rec int31_in_range_aux s ~min ~max =
    let r = Int64.to_int (next s) land 0x7FFF_FFFF in (* WRONG *)
    if r < min || r > max then int31_in_range_aux s ~min ~max else r
```
]

xavierleroy reviewed (changes requested):
Something is wrong in `int31_in_range_aux` and more generally in all this fancy footwork for 31/32/63 bit compatibility.  Time to go back to the drawing board.  In the meantime, it might be good to add tests to testsuite/tests/lib-random/chi2.ml : bugs like "it should draw numbers in [-1000,1000] but the result is always positive" are nicely caught by such statistical tests.

gmevel committed:
Random/test: add chi2 tests for int_in_range and similar

xavierleroy commented on testsuite/tests/lib-random/chi2.ml:
```diff
@@ -95,6 +111,175 @@ let _ =
+       (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 21);
+  let min_ = -214748364 in
+  let max_ = min_ + 0x3FFF_FFFF in
+  test "Random.int_in_range, range of length 2^30 (bits 0-7)"
+       (fun () -> Random.int_in_range ~min:min_ ~max:max_ - min_);
+  test "Random.int_in_range, range of length 2^30 (bits 22-29)"
+       (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 22);
+  let min_int31 = -0x4000_0000 in
+  let max_int31 = 0x3FFF_FFFF in
+  test "Random.int_in_range, full int31 range (bits 0-7)"
+      (fun () -> Random.int_in_range ~min:min_int31 ~max:max_int31);
+  test "Random.int_in_range, full int31 range (bits 23-30)"
+      (fun () -> (Random.int_in_range ~min:min_int31 ~max:max_int31) lsr 23);
+  test "Random.int_in_range, range of length 256*p < 2^30 (bits 0-7)"
+      (let p = 2097169 in (* prime < 2^22 *)
+       let min_ = -214748364 in
+       let max_ = min_ + (256 * p) - 1 in
+       fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
```
There's something fishy here.
- The quotient by p is uniformly distributed only if the interval is [n * p, m * p) for some n and m.  Otherwise, some  quotients appear more often than others.  Here, the low bound is not a multiple of p.  I'd rather use  `-128 * p` and `128 * p - 1` as bounds.
- OCaml's division is truncating division (rounds towards 0), not Euclidean division (rounds towards minus infinity), and this can bias the result.  Typically, if the low bound n * p is < 0 and the high bound m*p - 1 is > 0, the quotient n will appear almost never (instead of with probability 1/256) and the quotient 0 will appear too often (probability close to 2/256 instead of 1/256).  So, you should use Euclidean division here.

gmevel replied:
The test subtracts `min_` before dividing, so as to translate the interval to [0, 256*p). I think that addresses the issue as well.

Using `int_in_range` with an asymmetric interval was one of my goals, to avoid falling into what could conceivably be a special case.

xavierleroy replied:
Ah, right, I missed that.  In which case there should be no bias.

gmevel replied:
But now that I think of it, this is likely (?) wrong when `256 * p > max_int`. This happens in one of the later tests, depending on `Sys.int_size`.

In tests other than those with a division by p, It does not matter if the length of the interval is larger than `max_int` because, then, the subtraction `sample - min_` wraps modulo 2^int_size, and the chi2 test only cares for the lowest 8 bits.

gmevel replied:
I fixed this issue by using an unsigned division, i.e. by interpreting the dividend as an unsigned int. It turns out this operation is already provided by the standard library for `Int32` and `Int64`, I just had to implement it (partially, and not in the most efficient manner) for `int`:

```ocaml
(* Division of [x] by [y] where [x] is interpreted as an unsigned integer.
 * This code assumes [y >= 0]. *)
let udiv x y =
  if x >= 0 then x / y
  else
    let x' = x + min_int in
    let q = (x' / y) - (min_int / y)
    and r = (x' mod y) - (min_int mod y) in
    if r < y then q else q + 1
```

The biased chi2 test were passing nonetheless, I think, because the overflowing case is very unlikely (in these tests where `256 * p > max_int`, due to the way I chose it, `p` is such that `256 * p` is close to `max_int`, so it is very unlikely to actually draw a sample such that `sample - min_` overflows). If you’re happy with this solution, we can resolve this thread.

xavierleroy commented on testsuite/tests/lib-random/compat_32_64.ml:
```diff
@@ -0,0 +1,36 @@
+
+  let a = Array1.of_array Int64 C_layout [| 1L; 2L; 3L; 4L |] in
+  (* Violate abstraction of type Random.State.t to manipulate state directly *)
+  let r = (Obj.magic a : Random.State.t) in
```
Yurk!  Why?  Just use the default initial state of the default random generator, it's the same on all machines.

gmevel replied:
Done. (That part was copy-pasted from `testsuite/tests/lib-random/testvectors.ml` but indeed it is not useful for our test.)

xavierleroy reviewed (commented):
I read the code for `int_in_range` carefully, going so far as to draw the flowchart, and I think it is correct, although there may be ways to write this logic more concisely.  (I'll play with it later as I try to reintroduce the JS-of-OCaml special case that went away between 4.14 and 5.0.)

The chi-square tests may have biases, see below.

The 32/64 compatibility test is a good idea.  It seems to be failing on 32-bit x86, see Github CI logs, but I didn't investigate.

xavierleroy commented:
For the record, here is Euclidean division from truncating division :
```
let ediv a b =
  let q = a / b and r = a mod b in if r >= 0 then q else q - 1
```

gmevel commented on testsuite/tests/lib-random/chi2.ml:
```diff
@@ -95,6 +111,175 @@ let _ =
+         let max_ = min_ + (256 * p) - 1 in
+         fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
+  end;
+  test "Random.int_in_range, full int range (bits 0-7)"
+       (fun () -> Random.int_in_range ~min:min_int ~max:max_int);
+  test "Random.int_in_range, full int range (bits 30-37)"
+       (fun () -> (Random.int_in_range ~min:min_int ~max:max_int) lsr 30);
```
That test is what fails on i386. Obviously when `int` is 31 bits, sampling bits 30--37 is not uniform. This test should be moved in the conditional block `if Sys.int_size >= 63` (are 31, 32, 63 the only integer sizes we care about?).

gmevel committed:
Random/test: use default state in compat_32_64.ml

gmevel committed:
Random/test: fix coverage of chi2 tests

There are chi2 tests for `int_in_range` and similar, with intervals
whose length is comprised between 2^k and 2^(k+1) for some k; in some
tests, the length is 256*p, where p is some prime number such that
2^(k-8) < p < 2^(k-7).

Previously, p had been chosen to be the smallest prime number larger
than 2^(k-8), which made it very unlikely to draw samples larger than
min_ + 2^k; hence, the coverage of the test was very poor.

Now, p is chosen near 2^(k-8) * 3/2.

gmevel commented:
I just pushed a fix to the chi2 tests. Because of the way I had chosen constants, the coverage of some tests was very poor. In particular, it was not catching [the biased test spotted during review][msg-biased]. Now, if we replace unsigned divisions with signed divisions, “suspicious results” are reported as expected.

Quoting the commit message:

> There are chi2 tests for `int_in_range` and similar, with intervals
> whose length is comprised between 2^k and 2^(k+1) for some k; in some
> tests, the length is 256*p, where p is some prime number such that
> 2^(k-8) < p < 2^(k-7).
> 
> Previously, p had been chosen to be the smallest prime number larger
> than 2^(k-8), which made it very unlikely to draw samples larger than
> min_ + 2^k; hence, the coverage of the test was very poor.
> 
> Now, p is chosen near 2^(k-8) * 3/2.

Do you need anything from me?

[msg-biased]: https://github.com/ocaml/ocaml/pull/12459#discussion_r1286010750

xavierleroy committed:
Random: Refactor the implementation of [int] and of [int_in_range]

xavierleroy committed:
Random.{full_int,int_in_range}: add JavaScript / 64-bit OCaml compatibility

JS-of-OCaml uses JavaScript's 32-bit integers.  This commit makes sure
that we get the same results with JavaScript and with 64-bit OCaml.
For `Random.full_int` this was guaranteed in OCaml 4.14 but the code
to do so was wrongly removed in 5.0.

xavierleroy commented on stdlib/random.ml:
```diff
@@ -114,35 +114,112 @@ module State = struct
+      invalid_arg "Random.int_in_range"
+    (* When both bounds fit in 31 bits, we use [int31_in_range],
+     * so as to yield the same output on 32-bit and 64-bit platforms. *)
+    else if min < -0x4000_0000 || max > 0x3FFF_FFFF then
+      let span = max - min + 1 in
+      (* We use [int_in_range_aux] only if [max - min + 1] overflows;
+       * then, the interval between [min] and [max] covers at least half
+       * of the representable integers, so that it converges quickly
+       * (the probability of re-drawing is at most 1/2).
+       *
+       * When [max - min + 1] does not overflow, we use [full_int_aux],
+       * for 2 reasons:
+       *   - [int_in_range] then has a high probability of re-drawing,
+       *     so would be very slow;
+       *   - we thus guarantee that, when [min = 0], [int_in_range]
+       *     yields the same output as [full_int (max+1)].
+       *)
+      if span <= 0 then
+        int_in_range_aux s ~min ~max
+      else
+        min + full_int_aux s span
+    else
+      int31_in_range s ~min ~max
```
I think this code would be clearer if either
- `int31_in_range` was inlined here, resulting in 4 cases
- a function `full_int_in_range`, similar to `int31_in_range` was introduced and called here, resulting in 2 cases.

xavierleroy reviewed (approved):
At long last I was able to return to this code.  The current state looks good to me in terms of functionality and algorithms used.  The code is sometimes hard to read, see below for an example.

Cc @Octachron.

xavierleroy commented:
FYI: starting from this PR, I have two extensions in preparation:
- https://github.com/xavierleroy/ocaml/commit/a73ddddcfbfa9f5cb50f82942d3f6461e390d4ab refactors the code of `Random.{int,full_int,int_in_range}` to share more code between the 32-bit OCaml case and the general case.
- https://github.com/xavierleroy/ocaml/commit/5fd0768d031a364e6603a08ae270a56959314bd3 adds a second special case for compatibility with JavaScript and its 32-bit integers.  This compatibility is there in 4.14 for `Random.full_int` (PR #9489) but I wrongly removed it when reimplementing `Random` for 5.0.

gasche commented:
@gmevel we discussed this PR at today's triaging meeting; it seems that the ball is in your court. When you have time, could you consider Xavier's last-minute comments (and decide to change your code or not) so that we can merge?

xavierleroy commented:
I have submitted my changes as a pull request against this pull request: https://github.com/gmevel/ocaml/pull/1 .

We will also need a second reviewer from the core dev team, since this is a standard library change.

gmevel commented:
My apologies, I misunderstood that Xavier’s commits were part of a PR that would supersede this one, or that he would push his commits to this PR. Do you just need me to pull Xavier’s commits (I just did)?

gmevel commented:
+ [x] pulled Xavier’s commits
+ [x] rebased
+ [x] added a separate changelog entry for the JavaScript compatiblity (under the Standard Library section, feel free to move it to the Bugfixes section if you think that’s more appropriate)
  - [ ] reviewer to be filled in
+ [x] documented 5.2 as the first version with the new functions.

Should the JavaScript compatibility be mentioned in the doc? As the 32-bit OCaml / 64-bit OCaml already is.

gasche commented:
To clarify my request: Xavier made a "I think that the code would be clearer if" comment, which I was expecting you to react to in context of the present PR -- you could follow the suggestion, or decline to follow it, but make an explicit choice about it. He also mentioned two follow-up commit series that are going even further, and I did not expect you to integrate them in the present PR -- I think that Xavier could/should submit them as follow-up PRs once the present PR is merged.

If you prefer to revert your recent changes to un-include Xavier's follow-ups, that works for me. If you prefer to keep them, that also works for me -- I expect that you reviewed those changes yourself and believe they are correct.

I guess that I volunteered myself to be a second reviewer for this PR anyway.

gmevel commented:
> To clarify my request: Xavier made a "I think that the code would be clearer if" comment, which I was expecting you to react to in context of the present PR -- you could follow the suggestion, or decline to follow it, but make an explicit choice about it. He also mentioned two follow-up commit series that are going even further, and I did not expect you to integrate them in the present PR -- I think that Xavier could/should submit them as follow-up PRs once the present PR is merged.

I believe Xavier addressed his own reviewing remark (about refactoring) in his first commit (and I do agree with the said remark).

> If you prefer to revert your recent changes to un-include Xavier's follow-ups, that works for me. If you prefer to keep them, that also works for me -- I expect that you reviewed those changes yourself and believe they are correct.

I admit I haven’t really. I’ll try to find time, later, to read Xavier’s code more thoroughly.

xavierleroy commented:
Apologies if I wasn't clear enough, but by submitting my changes as a PR on this PR, I was hoping @gmevel would review my code (for a change :smiley: ) and suggest changes if needed.  Even though the chi2 test is getting really good at detecting bugs in the Random module, the code is nontrivial and contains some magic constants, so another pair of eyeballs is important.

I'm not sure yet how to test the JS-of-OCaml compatibility code, but maybe @Octachron knows.

gasche commented:
> I'm not sure yet how to test the JS-of-OCaml compatibility code, but maybe @Octachron knows.

The way I usually test js-of-ocaml code is by feeding it to [Try Ocaml](https://try.ocamlpro.com/). Unfortunately Try OCaml is currently based on OCaml 4.13, so the primitive `caml_lxm_next` is not available. Going this route would require creatively rebasing `int_in_range` on top of the previous PRNG, which is probably more work than checking the same thing using a local js_of_ocaml installation.

(It would be nice to get a 5.0 version of TryOCaml.)

dbuenzli commented:
> (It would be nice to get a 5.0 version of TryOCaml.)

It looks like the official OCaml playground is 5.0 https://ocaml.org/play

gmevel committed:
add changelog entry for JavaScript / 64-bit OCaml compat in Random

gmevel committed:
Random: fix typos in comments

also:
- reword "fit in X bits" to "fit in X-bit signed integers" (disambiguation)
- rename [int_in_range_gen] to [int_in_range_aux] (consistency with [int_aux])

gmevel committed:
Random: introduce {min,max}_int31 constants

gmevel committed:
Random/test: test JavaScript / 64-bit OCaml compat

gmevel committed:
Random/doc: document JavaScript / 64-bit OCaml compat

gmevel commented:
Ok, I reviewed Xavier’s code. The refactoring is a clear improvement, I couldn’t find a bug and all tests pass on my 64-bit installation, including the chi2 tests.

My only concern: I couldn’t wrap my head again around the math for the test in `int_aux` (despite that it is me who had written the explaining comment!), but since it is a direct generalization of the existing code, I assume it is correct. At least I can tell it does not overflow.

Since I had merged Xavier’s commit eagerly, I addressed my reviewing remarks myself:

- [x] Renamed `int_in_range_gen` to `int_in_range_aux` for consistency with `int_aux`.
- [x] Fixed a few typos in comments.

Further changes:

- [x] Refactored literal constants into named constants `min_int31` and `max_int31`.
- [x] Documented the compatibility between JavaScript and 64-bit OCaml.
- [x] Added a test for the said compatibility. Because the output would differ with 32-bit OCaml, we branch on `Sys.int_size`, and the expected output is not in the `*.reference` file but rather it is embedded in the source code.

Is JavaScript among the tested platforms though?

gasche commented:
> Is JavaScript among the tested platforms though?

This is a good question because the answer is "no", unfortunately.

Octachron commented:
For the javascript side, I was thinking of testing it in the https://github.com/ocaml/stdlib-random version of Random (which also has a OCaml-only version of the OCaml 5 PRNG).

gmevel commented:
> > Is JavaScript among the tested platforms though?
> 
> This is a good question because the answer is "no", unfortunately.

Then what shall I do with the test I just added? Revert it since it is void of object, or keep it for the future where platforms with >=32-bit integers might be added to the tested platforms (JavaScript, or an hypothetical implementation with 53-bit integers embedded in `float`, or 128-bit OCaml…)?

xavierleroy commented:
I used the "playground" at https://ocaml.org/play (based on JS-of-OCaml) to run the compat_32_64 test with a cut-down implementation of Random taken from this PR.  The output is as expected.

xavierleroy committed:
Let's not introduce typos in comments

xavierleroy commented:
As far as I can say, this PR is good for merging.  (Plus, I'm tired of reviewing it.)  A second approval from a core developer is required, however.

gasche commented:
Would @damiendoligez or @Ekdohibs be interested in reviewing this PR?

Octachron commented on stdlib/random.ml:
```diff
@@ -114,35 +114,107 @@ module State = struct
+     We must have [-2{^nbits - 1} <= min <= max < 2{^nbits - 1}].
+     Moreover, for the iteration to converge quickly, the interval
+     [[min, max]] should have width at least [2{^nbits - 1}]. *)
+  let rec int_in_range_alt s ~min ~max ~nbits =
```
Should we document that near the `2^{nbits - 1}` limit, we have a quite high standard deviation in the number of samples (around `2+ε`)?
The suffix `_alt` is also quite unclear, and it is not reused for the other variants of those functions: Maybe `int_in_large_range` would be more readable?

xavierleroy replied:
OK, done (7ebfbe4f89).

xavierleroy committed:
Cleaning up the random.ml implementation

Rename `int_in_range_aux` to `int_in_large_range`.
Comment on the expected number of draws.

xavierleroy committed:
Update changes for 12459

Octachron reviewed (approved):
The two rejection samplings are correct and restore the OS-independence of integer sampling.

Playing around with nodes confirmed that the implementation works and is testable with the `js_of_ocaml` backend (even if I am now tempted to generalize a bit the chi2 testing).

Octachron commented:
And in term of API, the `int_in_range` functions are certainly subtle enough to warrant inclusion in the `Random` module.

gasche commented:
Great! Thanks @Octachron, this should now be ready to merge.

The last remaining issue is the commit history. It is not so clean right now and I wonder what we would want to do about it.

xavierleroy commented:
> The last remaining issue is the commit history. It is not so clean right now and I wonder what we would want to do about it.

Squash!  With a killer commit message that I'm about to write.
